### PR TITLE
Smarter metadata fetching: topology

### PR DESCRIPTION
**Pull request by Lorak-mmk**
Based on https://github.com/scylladb/scylla-rust-driver/pull/1849

This PR changes the behavior on TOPOLOGY_CHANGE events. Previously, full metadata was fetched on each such event.
This PR now introduces a partial fetch, so only the topology (`system.local` and `system.peers`) is fetched.

Additionally, I added the exact same fetch on STATUS_CHANGE. This is because nodes `rpc_address` can change with just a restart. If we get rid of a periodic topology fetch, we would miss such changes.

Last commit introduces a small optimization: if the new topology is identical, there is no reason to publish new ClusterState.

Ref: https://github.com/scylladb/scylla-rust-driver/issues/1280

## Pre-review checklist

- [x] I have split my patch into logically separate commits.
- [x] All commit messages clearly explain what they change and why.
- [x] I added relevant tests for new features and bug fixes.
- [x] All commits compile, pass static checks and pass test.
- [x] PR description sums up the changes and reasons why they should be introduced.
- [x] I have provided docstrings for the public items that I want to introduce.
- [ ] ~~I have adjusted the documentation in `./docs/source/`.~~
- [x] I added appropriate `Fixes:` annotations to PR description.

### CODE REVIEW

**Comment by Lorak-mmk** [comment]
@coderabbitai review

File: scylla/src/cluster/metadata/fetching.rs [hunk 1]
```diff
@@ -192,22 +192,51 @@ impl ControlConnection {
...
         }
 
-        Ok(TopologyUpdateGuard::new(Metadata {
-            peers,
-            keyspaces,
-            cluster_name,
-            client_routes,
-        }))
+        Ok(())
     }
```
Comment (wprzytula) [inline scylla/src/cluster/metadata/fetching.rs:240]: ♻️ Extraction of `validate_peers` from `query_metadata` could come as a pre-commit.

File: scylla/src/cluster/metadata/update.rs [hunk 2]
```diff
@@ -61,6 +60,12 @@ pub(in super::super) struct PartialMetadataChanges {
     /// Partial client-routes snapshots fetched in response to
     /// CLIENT_ROUTES_CHANGE events.
     pub(in super::super) client_routes_updates: Option<ClientRoutesUpdate>,
+    /// The peer list fetched in response to TOPOLOGY_CHANGE events.
+    ///
+    /// Not partial within its own aspect: a topology fetch always reads the
+    /// whole peer list, so this replaces the topology of the state it is
+    /// applied to.
+    pub(in super::super) peers: Option<Vec<Peer>>,
```
Comment (wprzytula) [inline scylla/src/cluster/metadata/update.rs:68]: ❓ But why? If we only fetch topology in reaction to NODE_ADDED / NODE_REMOVED events, then wouldn't it be enough to fetch topology for that given node?
Perhaps the problem lies in identification: the events carry an IP address, while we identify nodes by host ID?

File: scylla/tests/integration/utils.rs [hunk 3]
```diff
@@ -727,3 +726,27 @@ pub(crate) fn inject_status_change_down(running_proxy: &RunningProxy, addr: Sock
...
+    );
+}
+
+/// Injects a forged `TOPOLOGY_CHANGE NEW_NODE` event for `addr`.
+pub(crate) fn inject_topology_change_new_node(running_proxy: &RunningProxy, addr: SocketAddr) {
+    inject_event_to_ccs(
+        running_proxy,
+        cluster_change_event_body("TOPOLOGY_CHANGE", "NEW_NODE", addr),
+    );
+}
```
Comment (wprzytula) [inline scylla/tests/integration/utils.rs:752]: ❓ What about `REMOVED_NODE`?

File: scylla/src/cluster/worker.rs
Comment (wprzytula) [inline scylla/src/cluster/worker.rs]: 🏕️ `wheter` -> `whether`

**Comment by Lorak-mmk** [comment]
Rebased on main.

File: scylla/src/cluster/metadata/update.rs (hunk 2 above)
Comment (Lorak-mmk) [reply scylla/src/cluster/metadata/update.rs:68]: Yes. I felt like fetching partial topology would open a can of edge-cases, for a small benefit.

File: scylla/tests/integration/utils.rs (hunk 3 above)
Comment (Lorak-mmk) [reply scylla/tests/integration/utils.rs:752]: I can add it if you want. I just thought the value of such test is low (we handle all topology change events the same way). For status change it had more sense because we do different things on up and down.

**Comment by Lorak-mmk** [comment]
- Addressed review comments
- Dropped the optimization that checked duplicate topology. I think we can do such things later, right now I'm aiming for a simple working solution.

**Comment by Lorak-mmk** [comment]
@coderabbitai full review

**Review by wprzytula — APPROVED** [review]

**Comment by Lorak-mmk** [comment]
Rebased on main

**Review by wprzytula — APPROVED** [review]
